dashboard/app/controllers/activities_controller.rb

```diff
@@ -113,29 +160,60 @@ def milestone_logger
     @@milestone_logger ||= Logger.new("#{Rails.root}/log/milestone.log")
   end
 
-  def track_progress_for_user
-    authorize! :create, Activity
-    authorize! :create, UserLevel
+  def track_level_solution(level:, program:, image:)
+    # Validate sharablility of program text and create a level source for it
+    if program
+      begin
+        share_failure = find_share_failure(program)
+      rescue OpenURI::HTTPError => share_checking_error
+        # If WebPurify fails, the program will be allowed
+      end
 
-    test_result = params[:testResult].to_i
-    solved = ('true' == params[:result])
+      level_source = LevelSource.find_identical_or_create(level, program) unless share_failure
+      slog(
+        tag: 'share_checking_error',
+        error: "#{share_checking_error}",
+        level_source_id: level_source.id
+      ) if share_checking_error && level_source
+    end
 
-    lines = params[:lines].to_i
+    # Store the image only if the image is set, and the image has not been saved
+    if image && level_source
```

Comment:
this is slightly changed compared to the previous code, as it checks against `level_source` instead of `level_source.try(:id)`. The latter will be `false` if the LevelSource is created but not yet saved (e.g., via the `find_identical_or_create` on L172 above). Not sure if this was an intentional change but if so it would be worth testing that the new logic behaves as expected on this edge case